File: server/cloud/gmail_oauth.py
"""Gmail OAuth (web flow) and encrypted token storage."""

from __future__ import annotations

import base64
import hashlib
import json
import os
import secrets
from typing import Any
from urllib.parse import urlencode

import httpx
from cryptography.fernet import Fernet, InvalidToken
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build

from server.cloud.config import (
    get_encryption_key,
    get_google_oauth_client_id,
    get_google_oauth_client_secret,
    get_google_oauth_redirect_uri,
)
from server.cloud.pg import execute, fetch_one
from server.services.gmail_client import SCOPES, GmailNotConfiguredError

GOOGLE_AUTH_URL = "https://accounts.google.com/o/oauth2/v2/auth"
GOOGLE_TOKEN_URL = "https://oauth2.googleapis.com/token"


def _fernet() -> Fernet:
    key = get_encryption_key()
    if not key:
        raise RuntimeError("TOKEN_ENCRYPTION_KEY is required for Gmail OAuth in cloud mode")
    # Fernet needs 32 url-safe base64-encoded bytes
    if len(key) == 44 and key.endswith("="):
        return Fernet(key.encode())
    derived = base64.urlsafe_b64encode(hashlib.sha256(key.encode()).digest())
    return Fernet(derived)


def encrypt_credentials(creds_json: str) -> str:
    return _fernet().encrypt(creds_json.encode()).decode()


def decrypt_credentials(encrypted: str) -> str:
    try:
        return _fernet().decrypt(encrypted.encode()).decode()
    except InvalidToken as exc:
        raise GmailNotConfiguredError("Invalid stored Gmail credentials") from exc


def oauth_configured() -> bool:
    return bool(
        get_google_oauth_client_id()
        and get_google_oauth_client_secret()
        and get_google_oauth_redirect_uri()
    )


def build_authorize_url(org_id: str, user_id: str) -> str:
    client_id = get_google_oauth_client_id()
    redirect_uri = get_google_oauth_redirect_uri()
    if not client_id or not redirect_uri:
        raise GmailNotConfiguredError("Google OAuth env vars not configured")
    state = secrets.token_urlsafe(32)
    execute(
        """
        INSERT INTO gmail_oauth_states (state, organization_id, user_id, created_at)
        VALUES (%s, %s, %s, now())
        ON CONFLICT (state) DO NOTHING
        """,
        (state, org_id, user_id),
    )
    params = {
        "client_id": client_id,
        "redirect_uri": redirect_uri,
        "response_type": "code",
        "scope": " ".join(SCOPES),
        "access_type": "offline",
        "prompt": "consent",
        "state": state,
    }
    return f"{GOOGLE_AUTH_URL}?{urlencode(params)}"


def exchange_code(code: str, state: str) -> dict[str, Any]:
    row = fetch_one(
        "SELECT organization_id, user_id FROM gmail_oauth_states WHERE state = %s",
        (state,),
    )
    if not row:
        raise ValueError("Invalid OAuth state")
    execute("DELETE FROM gmail_oauth_states WHERE state = %s", (state,))

    client_id = get_google_oauth_client_id()
    client_secret = get_google_oauth_client_secret()
    redirect_uri = get_google_oauth_redirect_uri()
    resp = httpx.post(
        GOOGLE_TOKEN_URL,
        data={
            "code": code,
            "client_id": client_id,
            "client_secret": client_secret,
            "redirect_uri": redirect_uri,
            "grant_type": "authorization_code",
        },
        timeout=30.0,
    )
    resp.raise_for_status()
    token_data = resp.json()

    creds = Credentials(
        token=token_data.get("access_token"),
        refresh_token=token_data.get("refresh_token"),
        token_uri=GOOGLE_TOKEN_URL,
        client_id=client_id,
        client_secret=client_secret,
        scopes=SCOPES,
    )
    service = build("gmail", "v1", credentials=creds)
    profile = service.users().getProfile(userId="me").execute()
    email_address = profile.get("emailAddress", "")

    org_id = str(row["organization_id"])
    user_id = str(row["user_id"])
    encrypted = encrypt_credentials(creds.to_json())

    execute(
        """
        INSERT INTO gmail_connections (
            organization_id, user_id, email_address, credentials_encrypted, scopes
        ) VALUES (%s, %s, %s, %s, %s)
        ON CONFLICT (organization_id, user_id) DO UPDATE SET
            email_address = EXCLUDED.email_address,
            credentials_encrypted = EXCLUDED.credentials_encrypted,
            scopes = EXCLUDED.scopes,
            updated_at = now()
        """,
        (org_id, user_id, email_address, encrypted, SCOPES),
    )
    return {"email_address": email_address, "organization_id": org_id}


def get_connection(org_id: str, user_id: str) -> dict[str, Any] | None:
    return fetch_one(
        """
        SELECT email_address, credentials_encrypted, connected_at, updated_at
        FROM gmail_connections
        WHERE organization_id = %s AND user_id = %s
        """,
        (org_id, user_id),
    )


def delete_connection(org_id: str, user_id: str) -> bool:
    from server.cloud.pg import get_conn

    with get_conn() as conn:
        cur = conn.execute(
            "DELETE FROM gmail_connections WHERE organization_id = %s AND user_id = %s",
            (org_id, user_id),
        )
        return cur.rowcount > 0


def get_gmail_service_for_user(org_id: str, user_id: str):
    conn_row = get_connection(org_id, user_id)
    if not conn_row:
        raise GmailNotConfiguredError("Gmail not connected — connect in Settings")
    creds_json = decrypt_credentials(conn_row["credentials_encrypted"])
    creds = Credentials.from_authorized_user_info(json.loads(creds_json), SCOPES)
    if not creds.valid:
        if creds.expired and creds.refresh_token:
            creds.refresh(Request())
            encrypted = encrypt_credentials(creds.to_json())
            execute(
                """
                UPDATE gmail_connections SET credentials_encrypted = %s, updated_at = now()
                WHERE organization_id = %s AND user_id = %s
                """,
                (encrypted, org_id, user_id),
            )
        else:
            raise GmailNotConfiguredError("Gmail token expired — reconnect in Settings")
    return build("gmail", "v1", credentials=creds)


def gmail_status(org_id: str, user_id: str) -> dict[str, Any]:
    if not oauth_configured():
        return {"ready": False, "error": "Google OAuth not configured on server"}
    conn = get_connection(org_id, user_id)
    if not conn:
        return {"ready": False, "error": "Gmail not connected", "connect_available": True}
    try:
        service = get_gmail_service_for_user(org_id, user_id)
        from server.services.gmail_client import get_sender_email
        email = get_sender_email(service)
        return {"ready": True, "from_email": email, "connected_at": str(conn.get("connected_at", ""))}
    except Exception as exc:  # noqa: BLE001
        return {"ready": False, "error": str(exc), "connect_available": True}

File: src/server/gmail-oauth.ts
import { randomBytes } from 'node:crypto'
import { and, eq } from 'drizzle-orm'
import { google } from 'googleapis'
import { getDb } from '#/db'
import { gmailConnections, gmailOauthStates } from '#/db/schema'
import { decryptText, encryptText } from '#/server/storage'

const SCOPES = [
  'https://www.googleapis.com/auth/gmail.readonly',
  'https://www.googleapis.com/auth/gmail.send',
]
const GOOGLE_AUTH_URL = 'https://accounts.google.com/o/oauth2/v2/auth'
const GOOGLE_TOKEN_URL = 'https://oauth2.googleapis.com/token'

function oauthConfig() {
  return {
    clientId: process.env.GOOGLE_OAUTH_CLIENT_ID ?? '',
    clientSecret: process.env.GOOGLE_OAUTH_CLIENT_SECRET ?? '',
    redirectUri: process.env.GOOGLE_OAUTH_REDIRECT_URI ?? '',
  }
}

export function oauthConfigured() {
  const c = oauthConfig()
  return Boolean(c.clientId && c.clientSecret && c.redirectUri)
}

export async function buildAuthorizeUrl(orgId: string, userId: string) {
  const { clientId, redirectUri } = oauthConfig()
  if (!clientId || !redirectUri) throw new Error('Google OAuth env vars not configured')
  const state = randomBytes(24).toString('base64url')
  const db = getDb()
  await db.insert(gmailOauthStates).values({ state, organizationId: orgId, userId })
  const params = new URLSearchParams({
    client_id: clientId,
    redirect_uri: redirectUri,
    response_type: 'code',
    scope: SCOPES.join(' '),
    access_type: 'offline',
    prompt: 'consent',
    state,
  })
  return `${GOOGLE_AUTH_URL}?${params}`
}

export async function exchangeCode(code: string, state: string) {
  const db = getDb()
  const row = await db.query.gmailOauthStates.findFirst({ where: eq(gmailOauthStates.state, state) })
  if (!row) throw new Error('Invalid OAuth state')
  await db.delete(gmailOauthStates).where(eq(gmailOauthStates.state, state))

  const { clientId, clientSecret, redirectUri } = oauthConfig()
  const tokenRes = await fetch(GOOGLE_TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      code,
      client_id: clientId,
      client_secret: clientSecret,
      redirect_uri: redirectUri,
      grant_type: 'authorization_code',
    }),
  })
  if (!tokenRes.ok) throw new Error('Token exchange failed')
  const tokenData = (await tokenRes.json()) as Record<string, unknown>

  const oauth2 = new google.auth.OAuth2(clientId, clientSecret, redirectUri)
  oauth2.setCredentials({
    access_token: tokenData.access_token as string,
    refresh_token: tokenData.refresh_token as string,
    scope: SCOPES.join(' '),
    token_type: 'Bearer',
  })
  const gmail = google.gmail({ version: 'v1', auth: oauth2 })
  const profile = await gmail.users.getProfile({ userId: 'me' })
  const emailAddress = profile.data.emailAddress ?? ''

  const credsJson = JSON.stringify({
    access_token: tokenData.access_token,
    refresh_token: tokenData.refresh_token,
    scope: SCOPES.join(' '),
    token_type: 'Bearer',
    client_id: clientId,
    client_secret: clientSecret,
  })

  await db
    .insert(gmailConnections)
    .values({
      organizationId: row.organizationId,
      userId: row.userId,
      emailAddress,
      credentialsEncrypted: encryptText(credsJson),
      scopes: SCOPES,
    })
    .onConflictDoUpdate({
      target: [gmailConnections.organizationId, gmailConnections.userId],
      set: {
        emailAddress,
        credentialsEncrypted: encryptText(credsJson),
        scopes: SCOPES,
        updatedAt: new Date(),
      },
    })

  return { email_address: emailAddress, organization_id: row.organizationId }
}

export async function getConnection(orgId: string, userId: string) {
  const db = getDb()
  return db.query.gmailConnections.findFirst({
    where: and(eq(gmailConnections.organizationId, orgId), eq(gmailConnections.userId, userId)),
  })
}

export async function deleteConnection(orgId: string, userId: string) {
  const db = getDb()
  const result = await db
    .delete(gmailConnections)
    .where(and(eq(gmailConnections.organizationId, orgId), eq(gmailConnections.userId, userId)))
    .returning({ id: gmailConnections.id })
  return result.length > 0
}

export async function getGmailServiceForUser(orgId: string, userId: string) {
  const conn = await getConnection(orgId, userId)
  if (!conn) throw new Error('Gmail not connected — connect in Settings')
  const credsJson = JSON.parse(decryptText(conn.credentialsEncrypted)) as Record<string, string>
  const { clientId, clientSecret, redirectUri } = oauthConfig()
  const oauth2 = new google.auth.OAuth2(clientId, clientSecret, redirectUri)
  oauth2.setCredentials({
    access_token: credsJson.access_token,
    refresh_token: credsJson.refresh_token,
    scope: credsJson.scope,
    token_type: credsJson.token_type,
  })
  return google.gmail({ version: 'v1', auth: oauth2 })
}

export async function gmailStatus(orgId: string, userId: string) {
  if (!oauthConfigured()) return { ready: false, error: 'Google OAuth not configured on server' }
  const conn = await getConnection(orgId, userId)
  if (!conn) return { ready: false, error: 'Gmail not connected', connect_available: true }
  try {
    const gmail = await getGmailServiceForUser(orgId, userId)
    const profile = await gmail.users.getProfile({ userId: 'me' })
    return {
      ready: true,
      from_email: profile.data.emailAddress,
      connected_at: conn.connectedAt.toISOString(),
    }
  } catch (e) {
    return { ready: false, error: String(e), connect_available: true }
  }
}

export { SCOPES }
